source/tests/universal/common/cases/descriptor/utils.py

```diff
@@ -93,3 +98,133 @@ def test_exclude_types(
+        mapping_device = self.mapping
+        # type_map for data and exclude_types
+        original_type_map = ["O", "H"]
+        full_type_map_test = [
+            "H",
+            "He",
+            "Li",
+            "Be",
+            "B",
+            "C",
+            "N",
+            "O",
+            "F",
+            "Ne",
+            "Na",
+            "Mg",
+            "Al",
+            "Si",
+            "P",
+            "S",
+            "Cl",
+            "Ar",
+        ]  # 18 elements
+        for old_tm, new_tm, em, econf in itertools.product(
+            [
+                deepcopy(full_type_map_test),  # 18 elements
+                deepcopy(
+                    full_type_map_test[:16]
+                ),  # 16 elements, double of tebd default first dim
+                deepcopy(full_type_map_test[:8]),  # 8 elements, tebd default first dim
+                ["H", "O"],  # slimmed types
+            ],  # old_type_map
+            [
+                deepcopy(full_type_map_test),  # 18 elements
+                deepcopy(
+                    full_type_map_test[:16]
+                ),  # 16 elements, double of tebd default first dim
+                deepcopy(full_type_map_test[:8]),  # 8 elements, tebd default first dim
+                ["H", "O"],  # slimmed types
+            ],  # new_type_map
+            [[], [[0, 1]], [[1, 1]]],  # exclude_types for original_type_map
+            [False, True],  # use_econf_tebd
+        ):
+            if len(old_tm) >= len(new_tm):
+                continue
+            # use shuffled type_map
+            shuffle(old_tm)
+            shuffle(new_tm)
+            old_tm_index = np.array(
+                [old_tm.index(i) for i in original_type_map], dtype=np.int32
+            )
+            new_tm_index = np.array(
+                [new_tm.index(i) for i in original_type_map], dtype=np.int32
+            )
+            old_tm_em = remap_exclude_types(em, original_type_map, old_tm)
+            old_tm_input = update_input_type_map(self.input_dict, old_tm)
+            old_tm_input = update_input_use_econf_tebd(old_tm_input, econf)
+            old_tm_input = update_input_exclude_types(old_tm_input, old_tm_em)
+            old_tm_module = self.module_class(**old_tm_input)
+            old_tm_dd = self.forward_wrapper(old_tm_module)
+            rd_old_tm, _, _, _, sw_old_tm = old_tm_dd(
+                coord_ext_device,
+                old_tm_index[atype_ext_device],
+                nlist_device,
+                mapping=mapping_device,
+            )
+            old_tm_module.change_type_map(new_tm)
+            new_tm_dd = self.forward_wrapper(old_tm_module)
+            rd_new_tm, _, _, _, sw_new_tm = new_tm_dd(
+                coord_ext_device,
+                new_tm_index[atype_ext_device],
+                nlist_device,
+                mapping=mapping_device,
+            )
+            np.testing.assert_allclose(rd_old_tm, rd_new_tm)
```

Comment:
The ut does not consider the case if the type of the input atoms are belonging to the new type map but not belonging to the old type map. 
In this case we do not test the consistency between the old and the new descriptors, but check if the stats for the new types are setup correctly.